core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/DedupAggregator.java

```diff
@@ -0,0 +1,187 @@
+  private List<String> uniqueFields;
+  private DedupConfig.DedupFunctionInfo filterFunction;
+
+  public DedupAggregator(DedupConfig dedupConfig) {
+    super(dedupConfig.numPartitions);
+    this.dedupConfig = dedupConfig;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    List<String> uniqueFields = dedupConfig.getUniqueFields();
+    DedupConfig.DedupFunctionInfo functionInfo = dedupConfig.getFilter();
+    if (functionInfo != null) {
+      // Invoke to validate whether the function used is supported
+      functionInfo.getSelectionFunction(null);
+    }
+
+    StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
+    Schema inputSchema = stageConfigurer.getInputSchema();
+    // if null, the input schema is unknown, or its multiple schemas.
+    if (inputSchema == null) {
+      stageConfigurer.setOutputSchema(null);
+      return;
+    }
+
+    // otherwise, we have a constant input schema. Get the output schema and propagate the schema
+    Schema outputSchema = getOutputSchema(inputSchema);
+    validateSchema(outputSchema, uniqueFields, functionInfo);
+    stageConfigurer.setOutputSchema(outputSchema);
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    uniqueFields = dedupConfig.getUniqueFields();
+    filterFunction = dedupConfig.getFilter();
+  }
+
+  @Override
+  public void groupBy(StructuredRecord record, Emitter<StructuredRecord> emitter) throws Exception {
+    if (uniqueFields == null) {
+      emitter.emit(record);
+      return;
+    }
+
+    StructuredRecord.Builder builder = StructuredRecord.builder(getGroupKeySchema(record.getSchema()));
+    for (String fieldName : uniqueFields) {
+      builder.set(fieldName, record.get(fieldName));
+    }
+    emitter.emit(builder.build());
+  }
+
+  @Override
+  public void aggregate(StructuredRecord groupKey, Iterator<StructuredRecord> iterator,
+                        Emitter<StructuredRecord> emitter) throws Exception {
+    if (!iterator.hasNext()) {
+      return;
+    }
+
+    SelectionFunction selectionFunction;
+    if (filterFunction == null) {
+      emitter.emit(iterator.next());
+    } else {
+      StructuredRecord firstRecord = iterator.next();
+      Schema.Field firstField = firstRecord.getSchema().getField(filterFunction.getField());
+      selectionFunction = filterFunction.getSelectionFunction(firstField.getSchema());
+      selectionFunction.beginFunction();
+      selectionFunction.operateOn(firstRecord);
+
+      while (iterator.hasNext()) {
+        selectionFunction.operateOn(iterator.next());
+      }
+
+      selectionFunction.finishFunction();
+      List<StructuredRecord> outputRecords = selectionFunction.getSelectedRecords();
+      for (StructuredRecord outputRecord : outputRecords) {
+        Schema outputSchema = getOutputSchema(outputRecord.getSchema());
+        validateSchema(outputSchema, uniqueFields, filterFunction);
```

Comment:
unless I'm misunderstanding, I believe the schema can never be invalid at this point.  In order to get here, all unique fields had to have been present when grouping, and the function had to have worked.

If there is any validation needed, it should happen in the group method.